Wangscape/image/TileRearranger.cpp

```diff
@@ -0,0 +1,263 @@
+                                                       dual_tessellation.slice(corner_codes[c]),
+                                                       false);
+    }
+    mTileRearrangement.dualCorners.set_size(arma::size(mTileRearrangement.dualEdges));
+    mTileRearrangement.dualCorners = mTileRearrangement.dualEdges;
+    for (unsigned int c = 0; c < Corners; c++)
+    {
+        mTileRearrangement.dualCorners.slice(c) %= mTileRearrangement.dualEdges.slice((c + Corners - 1) % Corners);
+    }
+    mTileRearrangement.dualCorners += opposites;
+    mTileRearrangement.dualCorners.transform([](arma::u8 x) {return std::min(x, arma::u8(1)); });
+
+    mTileRearrangement.dualCorners = unpaddedStack(mTileRearrangement.dualCorners);
+    mTileRearrangement.dualEdges = unpaddedStack(mTileRearrangement.dualEdges);
+}
+
+void TileRearranger::calculateRearrangementParameters()
+{
+    std::array<sf::IntRect, Corners> corner_bboxes;
+    for (int c = 0; c < Corners; c++)
+    {
+        corner_bboxes[c] = regionBox(mTileRearrangement.basePartition.slice(c));
+    }
+    // Depends on Corners
+    mTileRearrangement.allOffsets = {{1, 1}, {0, 1}, {0, 0}, {1, 0}};
+    for (auto& offset : mTileRearrangement.allOffsets)
+    {
+        offset = offset.x() * mTileRearrangement.offsetA + offset.y() * mTileRearrangement.offsetB;
+    }
+    std::array<sf::IntRect, Corners> dual_corner_bboxes;
+    for (unsigned int c = 0; c < Corners; c++)
+    {
+        dual_corner_bboxes[c] = corner_bboxes[c];
+        dual_corner_bboxes[c].left += mTileRearrangement.allOffsets[c].x();
+        dual_corner_bboxes[c].top += mTileRearrangement.allOffsets[c].y();
+    }
+    auto dual_bbox = dual_corner_bboxes[0];
+    for (unsigned int c = 1; c < Corners; c++)
+    {
+        dual_bbox = boxUnion(dual_bbox, dual_corner_bboxes[c]);
+    }
+    mTileRearrangement.dualSize = {(unsigned int)dual_bbox.width, (unsigned int)dual_bbox.height};
+    mTileRearrangement.dualOffset = {dual_bbox.left, dual_bbox.top};
+}
+
+void TileRearranger::makeDual()
+{
+    mTileRearrangement.dualPartition.set_size(mTileRearrangement.dualSize.y(),
+                                              mTileRearrangement.dualSize.x(),
+                                              Corners);
+    mTileRearrangement.dualPartition.fill(0);
+    for (unsigned int c = 0; c < Corners; c++)
+    {
+        copyRegionBounded(mTileRearrangement.basePartition.slice(c),
+                          mTileRearrangement.dualPartition.slice(c),
+                          {0, 0},
+                          mTileRearrangement.allOffsets[c] - mTileRearrangement.dualOffset,
+                          IVec(mTileRearrangement.baseSize));
+    }
+    mTileRearrangement.dual = arma::sum(mTileRearrangement.dualPartition, 2).eval();
+}
+
+std::vector<IVec> TileRearranger::makeTessellationOffsets() const
+{
+    std::vector<IVec> offsets;
+    for (int y = 0; y < 3; y++)
+    {
+        for (int x = 0; x < 3; x++)
```

Comment:
Why `3`s? It must be named.

Reply:
This should really be `for (int y=-1; y<=1; y++)`, with `y` coefficients inside the loop. Same for `x`.